### ISSUE TITLE
Python 3 failure in TkAlAllInOneTool unit test

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#29337]
It appears that #29255 introduced a failure in TkAlAllInOneTool unit test python3 IBs
```
TESTING all-in-one tool ...
Requesting file information for '/MinimumBias/Run2017A-TkAlMinBias-PromptReco-v1/ALCARECO' from DAS... Done.
Traceback (most recent call last):
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_PY3_X_2020-03-29-2300/bin/slc7_amd64_gcc820/validateAlignments.py", line 762, in &lt;module&gt;
    main()
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_PY3_X_2020-03-29-2300/bin/slc7_amd64_gcc820/validateAlignments.py", line 735, in main
    jobs.extend( ValidationJob(validation, config, options, 1) )
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_PY3_X_2020-03-29-2300/bin/slc7_amd64_gcc820/validateAlignments.py", line 132, in __init__
    self.validation = self.__getValidation( self.valType, self.valName,
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_PY3_X_2020-03-29-2300/bin/slc7_amd64_gcc820/validateAlignments.py", line 173, in __getValidation
    Alignment( alignments.strip(), config ), config )
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_PY3_X_2020-03-29-2300/python/Alignment/OfflineValidation/TkAlAllInOneTool/alignment.py", line 57, in __init__
    self.conditions = self.__getConditions( config, section )
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_PY3_X_2020-03-29-2300/python/Alignment/OfflineValidation/TkAlAllInOneTool/alignment.py", line 252, in __getConditions
    self.__testDbExist(condition["connectString"], condition["tagName"])
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_PY3_X_2020-03-29-2300/python/Alignment/OfflineValidation/TkAlAllInOneTool/alignment.py", line 260, in __testDbExist
    elif "\n"+tagname not in conddb("--db", dbpath.split("sqlite_file:")[1], "listTags"):
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_PY3_X_2020-03-29-2300/python/Alignment/OfflineValidation/TkAlAllInOneTool/helperFunctions.py", line 224, in conddb
    with NamedTemporaryFile(bufsize=0) as sys.stdout:
TypeError: NamedTemporaryFile() got an unexpected keyword argument 'bufsize'
```

**Comment by makortel** [comment]
assign alca

**Comment by makortel** [comment]
FYI @mmusich

**Comment by mmusich** [comment]
It's extremly annoying that one gets to know about this several days after the fact. How can we enforce this is discovered at integration time?

**Comment by makortel** [comment]
> It's extremly annoying that one gets to know about this several days after the fact. 

I agree (and sorry for not spotting it earlier).

> How can we enforce this is discovered at integration time?

Basically it means that we run tests on PY3 IB at least for PRs that have "non-trivial" python changes. Extending the testing to all PRs or those that have any python changes could be automatized, but would increase the testing load and thus latency to test results.

@smuzaffar @silviodonato what do you think?

**Comment by smuzaffar** [comment]
One of the idea of having Integration builds (IBs) is to look for these errors. It was laziness on our side , we should have identified the issue earlier.

I agree that we should test a PR for every possible IB we have but as @makortel wrote it will increse the laod on testing system.

**Comment by mmusich** [comment]
Possible solution at https://github.com/cms-sw/cmssw/pull/29339

**Comment by davidlange6** [comment]
Does our python dependency checking actually catch this sort of error (eg, induced by an interface change in another package)?

On Mar 30, 2020, at 3:47 PM, Matti Kortelainen <notifications@github.com<mailto:notifications@github.com>> wrote:



It's extremly annoying that one gets to know about this several days after the fact.

I agree (and sorry for not spotting it earlier).

How can we enforce this is discovered at integration time?

Basically it means that we run tests on PY3 IB at least for PRs that have "non-trivial" python changes. Extending the testing to all PRs or those that have any python changes could be automatized, but would increase the testing load and thus latency to test results.

@smuzaffar<https://github.com/smuzaffar> @silviodonato<https://github.com/silviodonato> what do you think?

**Comment by smuzaffar** [comment]
yes , python dependency keep track of dependency between package configuration only. It does not keep track of dependency of python files from test directory though.

We do run `python3 -m compileall` for PRs but looks like python3 did not complain about https://github.com/cms-sw/cmssw/pull/29255#discussion_r400193112
